Patch the list command to include the new Platformify command.

vintorez opened the pull request:
We need to override the platform list command on the Go side, to be able to include the new command (platform project:init) in the list of commands.

This should include both platform list and platform list project.

Technical details

Proxy to the legacy list command using --format=json Parse the JSON response
Add the extra commands, if needed
Return the enhanced result in the correct format (the one the user asked for)

Commit:
Move legacy package to internal

Commit:
Update go.mod and go.sum with new dependencies

Commit:
Convert version, help and root commands to Cobra commands

Commit:
Update everything to new paths

Commit:
Fix setting PLATFORMSH_CLI_DEBUG when the debug flag is set

Commit:
Add completion command

Commit:
Add golangci-lint

Commit:
Optimize PHP builds on Linux

Commit:
Integrate Platformify command in to the CLI

Commit:
Fix flag parsing in root commmand

Commit:
Bump platformify version and add commit in version next

vintorez committed:
Patch the list command to include the new Platformify command.

We need to override the platform list command on the Go side, to be able to include the new command (platform project:init) in
the list of commands.

This should include both platform list and platform list project.

Technical details

Proxy to the legacy list command using --format=json
Parse the JSON response
Add the extra commands, if needed
Return the enhanced result in the correct format (the one the user asked for)

vintorez committed:
fix linter

vintorez committed:
get rid of xml format

akalipetis commented on commands/list.go:
```diff
@@ -0,0 +1,115 @@
+
+var ListCmd = &cobra.Command{
+	Use:   "list [flags] [namespace]",
+	Short: "List",
```
```suggestion
	Short: "Lists commands",
```

This is the current description

vintorez replied:
fixed

akalipetis commented on commands/list.go:
```diff
@@ -0,0 +1,115 @@
+
+		if !all {
+			list.RemoveHiddenCommands()
+		}
```
This should better be handled by the Legacy CLI itself - you can add `--all` or not in the arguments.

vintorez replied:
agreed 👍 
replaced the function with the `--all` flag.
but it's not working properly now.
we need to fix this https://github.com/platformsh/legacy-cli/issues/1258 first

akalipetis replied:
Let's expect this to work correctly and as soon as the Legacy CLI is fixed, this will be fixed as well.

We'll need a confirmation from Product though that it's OK to expose those hidden values in the default renderer as well (they're exposed now anyway).

akalipetis commented on commands/list_formatters.go:
```diff
@@ -0,0 +1,90 @@
+type MDFormatter struct{}
+
+func (f *MDFormatter) Format(_ *List) ([]byte, error) {
+	// TODO: implement the method
```
Adding a comment here not to forget this.

vintorez replied:
added MarkDown builder

akalipetis reviewed (changes requested):
Added a few comments on the PR itself, but the output has (at least) the following issues:

* JSON - seems like we have a lot more options now
* Raw - we are showing a lot of shortcuts that seem to be hidden

This might be related to the following things:

1. We use `--all` by default, while we could potentially avoid this and only use it if used by the user
2. There are a few fields in the `List` structure that should potentially be `omitempty`

The following diffs should only be related to the `platformify` command:

```console
diff <(./platform list --format=json | jq --sort-keys .) <(platform list --format=json | jq --sort-keys .)
diff <(./platform list) <(platform list)
```

Also, things like using `--all`, etc should work as well.

vintorez committed:
Create MarkDown builder, sort arguments and options, use --all flag

akalipetis commented on commands/list.go:
```diff
@@ -0,0 +1,111 @@
+				formatter = &TXTListFormatter{}
+			}
+		default:
+			debugLog("%s\n", color.RedString("Unsupported format \"%s\".", format))
```
This does not have the same format as the command itself, a a quick fix you can run `list` in the Legacy CLI using this wrong format to get the correct error.

vintorez replied:
fixed. thanks for the hint!

vintorez committed:
fix unsupported format error message

vintorez committed:
update min golang version to 1.19, because Platformify requires 1.19

vintorez committed:
Change MD output, fix hidden options

vintorez committed:
remove html-like tags and support colors for lables like DEPRECATED

vintorez commented:
@akalipetis it is ready for review but it works only with Legacy-CLI from this PR: https://github.com/platformsh/legacy-cli/pull/1259
I have added support for color labels like `DEPRECATED` or `ALPHA`. 
I also added string "sanitizer" for removing all html-like tags (such as `<comment></comment>` or `<info></info>`).
The bad news is that it uses regular expressions.
I think everything is fine for now, but I hope that in the future we will be able to get rid of the use of regular expressions.

vintorez committed:
remove extra empty line

vintorez commented:
closing this one.
see: https://github.com/platformsh/cli/pull/60
